lib/providers/portion/PortionProvider.ts

```diff
@@ -0,0 +1,255 @@
+   *
+   * @param sharedInfo shared quote request info across classic and dutch
+   * @param tokenOutAmount
+   *
+   * @return FullPortionPayload.portionAmount
+   * for exact in, this will be the QUOTE (not quote with gas adjusted) amount from router (tokenOut amount) multiply by the portion Bps divided by 10k basis point (10000).
+   * for exact out, this will be the user typed in amount (tokenOut amount) multiply by the portion Bps divided by 10k basis point (10000).
+   *
+   * portion amount will be unit-less, meaning amount scaled out by tokenOut's decimals
+   *
+   * @return FullPortionPayload.portionAdjustedTokenOutAmount
+   * for exact in, this will be the quote WITH GAS ADJUSTED (not quote) amount from router (tokenOut amount) minus the portion amount.
+   * for exact out, this will be the user typed in amount (tokenOut amount) plus the portion amount.
+   *
+   * portion adjusted token out amount will be unit-less, meaning amount scaled out by tokenOut's decimals
+   *
+   */
+  getPortionAdjustedQuote(
+    sharedInfo: QuoteRequestInfo,
+    quote: string,
+    quoteGasAdjusted: string,
+    tokenOutAmount: string
+  ): Promise<FullPortionPayload>;
+}
+
+export class DefaultPortionProvider implements PortionProvider {
+  constructor(private portionFetcher: PortionFetcher, private tokenFetcher: TokenFetcher) {}
+
+  async getPortion(sharedInfo: QuoteRequestInfo): Promise<GetPortionResponse> {
+    // we will need to call token fetcher to resolve the tokenIn and tokenOut
+    // there's no guarantee that the tokenIn and tokenOut are in the token address
+    // also the tokenIn and tokenOut can be native token
+    // portion service only accepts wrapped token address
+    const [resolvedTokenIn, resolvedTokenOut] = await this.resolveTokenInAndTokenOut(sharedInfo);
+
+    // portion service has no concept of native currency, need to pass in wrapped
+    const getPortionResponse =
+      resolvedTokenIn && resolvedTokenOut
+        ? await this.portionFetcher.getPortion(
+          sharedInfo.tokenInChainId,
+          resolvedTokenIn.wrapped.address,
+          sharedInfo.tokenOutChainId,
+          resolvedTokenOut.wrapped.address
+        )
+        : undefined;
+
+    return getPortionResponse ?? GET_NO_PORTION_RESPONSE;
+  }
+
+  async getPortionForTokenOut(sharedInfo: QuoteRequestInfo, tokenOutAmount: string): Promise<FullPortionPayload> {
+    // we will need to call token fetcher to resolve the tokenIn and tokenOut
+    // there's no guarantee that the tokenIn and tokenOut are in the token address
+    // also the tokenIn and tokenOut can be native token
+    // portion service only accepts wrapped token address
+    const [_, resolvedTokenOut] = await this.resolveTokenInAndTokenOut(sharedInfo);
+
+    const getPortionResponse = await this.getPortion(sharedInfo);
+
+    const tokenOutCurrencyAmount = resolvedTokenOut
+      ? CurrencyAmount.fromRawAmount(resolvedTokenOut, tokenOutAmount)
+      : undefined;
```

Comment:
like above, you could likely return early here if this `resolvedTokenOut` in undefined, I imagine all the fields in the response would be undefined anyways... and then you don't have to special case for undefinedness in the other functions in this class